verl/utils/reward_score/sandbox_fusion/utils.py

```diff
@@ -0,0 +1,378 @@
+                 result_status = -1 # Classify as unknown error
+        else:
+             # API returned an unknown top-level status
+             logger.warning(f"Unknown API status received: {api_status}")
+             metadata["status"] = f"unknown_api_status_{api_status}"
+             result_status = -1 # Default to -1
+    else: # api_response is None and no error_msg (Should not happen with current call_sandbox_api logic)
+        metadata["status"] = "unknown_api_state"
+        result_status = -1
+        logger.error(f"Case {case_index}: Unknown API state (no response and no error message).")
+    return result_status, metadata
+
+
+def check_correctness(sandbox_fusion_url: str, in_outs: Optional[dict], generation: str, timeout: int = DEFAULT_TIMEOUT, language: str = "python") -> Tuple[List[Any], List[Dict[str, Any]]]:
+    """
+    Checks the correctness of code generation using the remote sandbox API,
+    processing test cases concurrently.
+
+    Args:
+        sandbox_fusion_url: The URL of the sandbox fusion API.
+        in_outs: Dictionary containing "inputs" and "outputs" lists.
+        generation: The generated code string.
+        timeout: Timeout for each test case (compile and run share this timeout).
+        language: The programming language of the code.
+
+    Returns:
+        A tuple (results, metadata_list).
+        results: A list containing the test result for each input/output pair
+                 (True/False/-1 api/sandbox err, -2 runtime err, -3 timeout, -4 compile err).
+                 Results are ordered corresponding to the inputs.
+        metadata_list: A list containing metadata dictionaries for each test case,
+                       ordered corresponding to the inputs.
+    """
+    logger.info(f"Starting correctness check for generation.")
+
+    if not in_outs or "inputs" not in in_outs or "outputs" not in in_outs:
+        logger.warning(f"Invalid in_outs format provided.")
+        return [-1], [{"error": "Invalid input/output data"}]
+
+    inputs = in_outs["inputs"]
+    expected_outputs = in_outs["outputs"]
+    num_cases = len(inputs)
+    results = [None] * num_cases # Initialize with placeholders
+    metadata_list = [None] * num_cases # Initialize with placeholders
+
+    if num_cases == 0:
+        logger.warning("Empty inputs provided.")
+        return [], []
+
+    if len(inputs) != len(expected_outputs):
+         logger.warning(f"Mismatch between number of inputs ({len(inputs)}) and outputs ({len(expected_outputs)}).")
+         # Return error based on the number of inputs provided
+         return [-1] * num_cases, [{"error": "Input/output count mismatch", "case_index": i} for i in range(num_cases)]
+
+
+    first_compile_error_index = -1
+
+    # Use ThreadPoolExecutor for concurrent execution
+    # Adjust max_workers based on expected load and API limits
+    with concurrent.futures.ThreadPoolExecutor() as executor:
```

Comment:
For multi turn rollout, each chat completion may independently call check_correctness, and we may have thousands of chat completion running concurrently. So we better have a global ThreadPoolExecutor, or more ideally use asyncio to avoid invoke hundreds of thread.

A `SandboxFusionManager` may a better option for  thread pool/event loop, rate limit control, etc.
```python
class SandboxFusionManager:
    def submit(self, in_outs: Optional[dict], generation: str, ...):
        pass
```

Reply:
1. implemented a global concurrent limit to fusion api calls by using a thread semaphore
2. since prime.py calls compute_score method in subprocess concurrently, it can not work with asyncio in sandbox fusion. I will submit another version of sandbox for asyncio soon with SandboxFusionManager.